Isinglike

AlexanderIbrahim1 opened the pull request:
* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

This adds a function `isinglike_analysis()`. It takes an `AnalogGate` instance that implements an Ising-like Hamiltonian, and returns a dictionary that maps the type of interaction (`XX`, `XY`, `YY`, etc.) to the corresponding coupling matrix. If the input is invalid, for any of the reasons given in the docstring, a `RuntimeError` is raised.

It also has an optional argument `allowed_terms` that allows users to choose which of the Ising-like interaction types should be allowed in the Hamiltonian.

* **What is the current behavior?** (You can also link to an open issue here)

this change addresses #59


* **What is the new behavior (if this is a feature change)?**



* **Does this PR introduce a breaking change?** (What changes might users need 
to make in their application due to this PR?)

There is no breaking change.


* **Other information**:

I have no idea how to update the docs.

I'm not entirely familiar with all the literature and terminology in this subfield, so I'm not 100% sure if the outputs (in the unit tests) are the intended outputs.

yhteoh commented:
Hello Alexander, thank you so much for the contribution, I see in your code that you implemented your own traversal structure for traversing the Operator AST and performing the analysis, we would appreciate it if you were to make use of our [compiler infrastructure](https://github.com/OpenQuantumDesign/oqd-compiler-infrastructure) that implements the walks and rules used to traverse and analyse the Operator AST respectively.

AlexanderIbrahim1 commented:
Hello yhteoh,

Thanks for the advice.

I've replaced the custom traversal structure with one from OQD's infrastructure.
I also fixed a bug where like terms (XX, YY, ZZ) were double counted.

Also, I initially submitted this PR before UnitaryHack started; do I need to resubmit the PR for it to count?
